File: server/services/apiConnector.js
const axios = require('axios');
const APIConfig = require('../models/APIConfig'); // Assuming path

// Basic logger
const logger = {
    info: (msg) => console.log(\`[ApiConnector INFO] \${msg}\`),
    warn: (msg) => console.warn(\`[ApiConnector WARN] \${msg}\`),
    error: (msg) => console.error(\`[ApiConnector ERROR] \${msg}\`)
};

class APIException extends Error {
    constructor(message, status) {
        super(message);
        this.name = "APIException";
        this.status = status;
    }
}

class ApiConnector {
    constructor(apiConfig) {
        if (!(apiConfig instanceof APIConfig)) {
            logger.warn('ApiConnector initialized with invalid APIConfig. Using default.');
            this.config = new APIConfig();
        } else {
            this.config = apiConfig;
        }
        this.debugCallback = (msg) => logger.info(msg); // Default debug callback
    }

    setDebugCallback(callback) {
        this.debugCallback = callback;
    }

    async callApi(messages, max_tokens = 2000, temperature = 0.7, use_planning_model = false) {
        const currentConfig = use_planning_model && this.config.use_planning_model
            ? {
                provider: this.config.planning_provider || this.config.provider,
                apiKey: this.config.planning_api_key || this.config.api_key,
                baseUrl: this.config.planning_base_url || this.config.base_url,
                model: this.config.planning_model || this.config.model,
              }
            : {
                provider: this.config.provider,
                apiKey: this.config.api_key,
                baseUrl: this.config.base_url,
                model: this.config.model,
              };

        if (!currentConfig.apiKey) {
            this.debugCallback('❌ API Key is missing.');
            throw new APIException("API Key is missing.", 401);
        }
        this.debugCallback(\`💡 Using model: \${currentConfig.model} from \${currentConfig.provider} at \${currentConfig.baseUrl}\`);

        let lastError = null;
        for (let attempt = 0; attempt < this.config.max_retries; attempt++) {
            try {
                this.debugCallback(\`📤 API call attempt \${attempt + 1}/\${this.config.max_retries}\`);
                let response;
                if (currentConfig.provider === 'openai' || currentConfig.provider === 'custom' || currentConfig.provider.includes('ollama') || currentConfig.provider.includes('lm-studio') || current_config.provider.includes('localai') || current_config.provider.includes('text-generation-webui') || current_config.provider.includes('vllm')) {
                    response = await this._callOpenAICompatAPI(messages, max_tokens, temperature, currentConfig);
                } else if (currentConfig.provider === 'anthropic') {
                    response = await this._callAnthropicAPI(messages, max_tokens, temperature, currentConfig);
                } else {
                    throw new APIException(\`Unsupported API provider: \${currentConfig.provider}\`, 400);
                }
                return response;
            } catch (e) {
                lastError = e;
                this.debugCallback(\`⚠️ API call attempt \${attempt + 1} failed: \${e.message}\`);
                if (e instanceof APIException && (e.status === 401 || e.status === 403 || e.status === 429)) {
                    // Don't retry on auth errors or rate limits immediately
                    throw e;
                }
                if (attempt === this.config.max_retries - 1) {
                    this.debugCallback('❌ API call failed after max retries.');
                    throw new APIException(\`API call failed after \${this.config.max_retries} retries: \${lastError.message}\`, lastError.status || 500);
                }
                // Wait a bit before retrying (optional, simple delay)
                await new Promise(resolve => setTimeout(resolve, 500 * (attempt + 1)));
            }
        }
        // Should not be reached if max_retries > 0
        throw lastError || new APIException('API call failed.', 500);
    }

    async _callOpenAICompatAPI(messages, max_tokens, temperature, config) {
        const headers = {
            "Authorization": \`Bearer \${config.apiKey}\`,
            "Content-Type": "application/json"
        };
        const body = {
            model: config.model,
            messages: messages,
            max_tokens: max_tokens,
            temperature: temperature
        };
        if (this.config.disable_thinking) { // Check global config for disable_thinking
            body.thinking = false;
        }

        try {
            const response = await axios.post(\`\${config.baseUrl}/chat/completions\`, body, {
                headers: headers,
                timeout: this.config.timeout * 1000 // axios timeout is in ms
            });

            if (response.data && response.data.choices && response.data.choices[0]) {
                return {
                    content: response.data.choices[0].message.content,
                    usage: response.data.usage,
                    model: response.data.model || config.model
                };
            } else {
                throw new APIException('Invalid response structure from OpenAI compatible API.', 500);
            }
        } catch (error) {
            if (error.response) {
                throw new APIException(\`OpenAI API Error: \${error.response.status} \${JSON.stringify(error.response.data)}\`, error.response.status);
            } else if (error.request) {
                throw new APIException('No response received from OpenAI API.', 504);
            } else {
                throw new APIException(\`Error setting up OpenAI API request: \${error.message}\`, 500);
            }
        }
    }

    async _callAnthropicAPI(messages, max_tokens, temperature, config) {
        const headers = {
            "x-api-key": config.apiKey,
            "Content-Type": "application/json",
            "anthropic-version": "2023-06-01"
        };

        let system_message = "";
        let user_messages = messages;
        if (messages && messages[0] && messages[0].role === "system") {
            system_message = messages[0].content;
            user_messages = messages.slice(1);
        }

        const body = {
            model: config.model,
            max_tokens: max_tokens,
            temperature: temperature,
            messages: user_messages,
        };
        if (system_message) {
            body.system = system_message;
        }

        try {
            const response = await axios.post(\`\${config.baseUrl}/messages\`, body, { // Anthropic base URL might not need /v1
                headers: headers,
                timeout: this.config.timeout * 1000
            });

            if (response.data && response.data.content && response.data.content[0]) {
                return {
                    content: response.data.content[0].text,
                    usage: response.data.usage,
                    model: response.data.model || config.model // Anthropic response might have model in `response.data.model` or similar
                };
            } else {
                throw new APIException('Invalid response structure from Anthropic API.', 500);
            }
        } catch (error) {
            if (error.response) {
                throw new APIException(\`Anthropic API Error: \${error.response.status} \${JSON.stringify(error.response.data)}\`, error.response.status);
            } else if (error.request) {
                throw new APIException('No response received from Anthropic API.', 504);
            } else {
                throw new APIException(\`Error setting up Anthropic API request: \${error.message}\`, 500);
            }
        }
    }
}

module.exports = { ApiConnector, APIException };

File: novel_writer/services/api_connector.py
"""
API連接器模組 - 從 novel_writer.py 重構抽取
處理所有LLM API調用邏輯，支持OpenAI、Anthropic和自訂API
"""

import requests
import logging
from typing import Dict, List, Callable

from ..models import APIConfig, APIException

# 配置日誌
logger = logging.getLogger(__name__)


class APIConnector:
    """LLM API連接器 - 重構版"""
    
    def __init__(self, config: APIConfig, debug_callback: Callable = None):
        self.config = config
        self.debug_callback = debug_callback or (lambda x: None)
        
    def call_api(self, messages: List[Dict], max_tokens: int = 2000, 
                temperature: float = 0.7, use_planning_model: bool = False) -> Dict:
        """調用LLM API with retry logic"""
        
        # 根據是否使用規劃模型選擇配置
        if use_planning_model and self.config.use_planning_model:
            provider = self.config.planning_provider
            api_key = self.config.planning_api_key or self.config.api_key
            base_url = self.config.planning_base_url
            model = self.config.planning_model
            self.debug_callback("💡 使用規劃模型進行API調用")
        else:
            provider = self.config.provider
            api_key = self.config.api_key
            base_url = self.config.base_url
            model = self.config.model
        
        for attempt in range(self.config.max_retries):
            try:
                logger.info(f"API調用嘗試 {attempt + 1}/{self.config.max_retries} (模型: {model})")
                
                if provider == "openai":
                    return self._call_openai_api(messages, max_tokens, temperature, api_key, base_url, model)
                elif provider == "anthropic":
                    return self._call_anthropic_api(messages, max_tokens, temperature, api_key, base_url, model)
                elif provider == "custom":
                    return self._call_custom_api(messages, max_tokens, temperature, api_key, base_url, model)
                else:
                    raise APIException(f"不支持的API提供商: {provider}")
                    
            except requests.exceptions.RequestException as e:
                logger.warning(f"API調用失敗 (嘗試 {attempt + 1}): {str(e)}")
                if attempt == self.config.max_retries - 1:
                    raise APIException(f"API調用失敗，已重試 {self.config.max_retries} 次: {str(e)}")
            except Exception as e:
                logger.error(f"API調用出現未預期錯誤: {str(e)}")
                raise APIException(f"API調用錯誤: {str(e)}")
    
    def _call_openai_api(self, messages: List[Dict], max_tokens: int, 
                        temperature: float, api_key: str, base_url: str, model: str) -> Dict:
        """調用OpenAI格式API"""
        headers = {
            "Authorization": f"Bearer {api_key}",
            "Content-Type": "application/json"
        }
        
        data = {
            "model": model,
            "messages": messages,
            "max_tokens": max_tokens,
            "temperature": temperature
        }
        
        # 如果啟用了關閉thinking，添加thinking參數
        if self.config.disable_thinking:
            data["thinking"] = False
        
        response = requests.post(
            f"{base_url}/chat/completions",
            headers=headers,
            json=data,
            timeout=self.config.timeout
        )
        
        if response.status_code == 200:
            result = response.json()
            return {
                "content": result["choices"][0]["message"]["content"],
                "usage": result.get("usage", {}),
                "model": result.get("model", model)
            }
        else:
            raise APIException(f"API調用失敗: {response.status_code} {response.text}")
    
    def _call_anthropic_api(self, messages: List[Dict], max_tokens: int, 
                           temperature: float, api_key: str, base_url: str, model: str) -> Dict:
        """調用Anthropic API"""
        headers = {
            "x-api-key": api_key,
            "Content-Type": "application/json",
            "anthropic-version": "2023-06-01"
        }
        
        # 轉換消息格式
        system_message = ""
        if messages and messages[0]["role"] == "system":
            system_message = messages[0]["content"]
            messages = messages[1:]
        
        data = {
            "model": model,
            "max_tokens": max_tokens,
            "temperature": temperature,
            "messages": messages
        }
        
        if system_message:
            data["system"] = system_message
        
        response = requests.post(
            f"{base_url}/messages",
            headers=headers,
            json=data,
            timeout=self.config.timeout
        )
        
        if response.status_code == 200:
            result = response.json()
            return {
                "content": result["content"][0]["text"],
                "usage": result.get("usage", {}),
                "model": result.get("model", model)
            }
        else:
            raise APIException(f"API調用失敗: {response.status_code} {response.text}")
    
    def _call_custom_api(self, messages: List[Dict], max_tokens: int, 
                        temperature: float, api_key: str, base_url: str, model: str) -> Dict:
        """調用自訂API"""
        return self._call_openai_api(messages, max_tokens, temperature, api_key, base_url, model)